drivers/nrf9160_gps/nrf9160_gps.c

```diff
@@ -0,0 +1,527 @@
+	}
+}
+
+static int init_thread(struct device *dev)
+{
+	struct gps_drv_data *drv_data = dev->driver_data;
+
+	k_thread_create(&drv_data->thread, drv_data->thread_stack,
+			K_THREAD_STACK_SIZEOF(drv_data->thread_stack),
+			(k_thread_entry_t)gps_thread, dev, NULL, NULL,
+			K_PRIO_PREEMPT(CONFIG_NRF9160_GPS_THREAD_PRIORITY),
+			0, 0);
+
+	return 0;
+}
+
+static int enable_gps(struct device *dev)
+{
+	int err;
+	char buf[50] = {0};
+	struct at_param_list at_response_list = {0};
+	u16_t gps_param_value, functional_mode;
+
+	err = at_cmd_write(AT_XSYSTEMMODE_REQUEST, buf, sizeof(buf), NULL);
+	if (err) {
+		LOG_ERR("Could not get modem's system mode");
+		return -EIO;
+	}
+
+	at_params_list_init(&at_response_list, AT_XSYSTEMMODE_PARAMS_COUNT);
+
+	err = at_parser_max_params_from_str(
+		&buf[sizeof(AT_XSYSTEMMODE_RESPONSE)],
+		&at_response_list,
+		AT_XSYSTEMMODE_PARAMS_COUNT);
+	if (err) {
+		LOG_ERR("Could not parse AT response, error: %d", err);
+		return err;
+	}
+
+	err = at_params_short_get(&at_response_list,
+				  AT_XSYSTEMMODE_GPS_PARAM_INDEX,
+				  &gps_param_value);
+	if (err) {
+		LOG_ERR("Could not get GPS mode state, error: %d", err);
+		return err;
+	}
+
+	if (gps_param_value != 1) {
+		char cmd[sizeof(AT_XSYSTEMMODE_PROTO)];
+		size_t len;
+		u16_t values[AT_XSYSTEMMODE_PARAMS_COUNT] = {0};
+
+		LOG_DBG("GPS mode is not enabled, attempting to enable it");
+
+		for (size_t i = 0; i < AT_XSYSTEMMODE_PARAMS_COUNT; i++) {
+			at_params_short_get(&at_response_list, i, &values[i]);
+		}
+
+		values[AT_XSYSTEMMODE_GPS_PARAM_INDEX] = 1;
+
+		len = snprintf(cmd, sizeof(cmd), AT_XSYSTEMMODE_PROTO,
+			       values[0], values[1], values[2], values[3]);
+
+		LOG_DBG("Sending AT command: %s", log_strdup(cmd));
+
+		err = at_cmd_write(cmd, NULL, 0, NULL);
+		if (err) {
+			LOG_ERR("Could not enable GPS mode, error: %d", err);
+			return err;
+		}
+	}
+
+	LOG_DBG("GPS mode is enabled");
+
+	err = at_cmd_write(AT_CFUN_REQUEST, buf, sizeof(buf), NULL);
+	if (err) {
+		LOG_ERR("Could not get functional mode, error: %d", err);
+		return err;
+	}
+
+	err = at_parser_max_params_from_str(
+		&buf[sizeof(AT_CFUN_RESPONSE)],
+		&at_response_list, 1);
+	if (err) {
+		LOG_ERR("Could not parse functional mode response, error: %d",
+			err);
+		return err;
+	}
+
+	err = at_params_short_get(&at_response_list, 0, &functional_mode);
+	if (err) {
+		LOG_ERR("Could not get value of functional mode, error: %d",
+			err);
+		return err;
+	}
+
+	LOG_DBG("Functional mode: %d", functional_mode);
+
+	if (functional_mode != FUNCTIONAL_MODE_ENABLED) {
+		LOG_DBG("Functional mode was %d, attemping to set to %d",
+			functional_mode, FUNCTIONAL_MODE_ENABLED);
+
+		err = at_cmd_write(AT_CFUN_1, NULL, 0, NULL);
+		if (err) {
+			LOG_ERR("Could not set functional mode to %d",
+				FUNCTIONAL_MODE_ENABLED);
+			return err;
+		}
+		LOG_DBG("Functional mode set to %d", FUNCTIONAL_MODE_ENABLED);
+	}
+
+	at_params_list_free(&at_response_list);
```

Comment:
There's multiple return paths in this function, which do not free the allocated list. I propose that each err has a "goto error" where the list is free'd.

Reply:
Good catch. Since the list should be freed either way, I went for a "clean exit" approach. See the latest commits.